final-project/static/scripts.js

```diff
@@ -6,23 +6,59 @@
 // Enforce strict js code conventions
 "use strict";
 
-// Render product on screen
-function displayProduct() {
+// Render the product on screen.
+$(function retrieveJSON() {
 
-  // TODO: create parameters for Flask.url_for() method
-  // TODO: use Flask.url_for("product", parameters) to generate JSON url
 
-  // TODO: create variable to store our JSON data
-    // (probably in jqXHR format)
+  // Get the JSON with $.getJSON() & Flask.url_for().
+    // Use Flask.url_for("product", parameters) to generate JSON url.
+  function getProductInfo (parameters, callback) {
+    $.getJSON(Flask.url_for("product"), parameters)
+    // A callback executed upon successful retrieval.
+    .done(function(data, textStatus, jqXHR) {
+      console.log(textStatus)
+      console.log(jqXHR)
+      callback(jqXHR)
+    })
+    // A callback executed upon failed retrieval.
+    .fail(function(data, textStatus, errorThrown) {
+      // Log an error to the browser.
+      console.log(errorThrown.toString());
+    });
+  }
 
-  // TODO: get the JSON with $.getJSON() & Flask.url_for()
-    // TODO: use callback to check for successful data retrieval
+  // Get the element id attribute when a product is clicked
+  $("#recently_added_products").click(function(event) {
```

Comment:
Remember that this is async, so that the order at which this code execute is per line. If something happens in a callback, it won't run until the async code finishes.

Reply:
You can take a look at ES6's `await` for better syntax, or you can chain together your callbacks/promises.

Reply:
Would you be up to walk me through what you envision for this code structure? I started to turn my AJAX into a Promise, but the code looks messier than before. Not to mention, it doesn't work yet. I'll push the current code now so you can see what I mean.

I read through Google's explanation of promises and Mozilla's.

Google: https://developers.google.com/web/fundamentals/primers/promises#bonus_round_promises_and_generators 

Mozilla: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises 

Looks like I need a promise before I can use await.